bulk_indexer_pool.go

```diff
@@ -0,0 +1,209 @@
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package docappender
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// BulkIndexerPool is a pool of BulkIndexer instances. It is designed to be
+// used in a concurrent environment where multiple goroutines may need to
+// acquire and release indexers.
+//
+// The pool allows a minimum number of BulkIndexers to be guaranteed per ID, a
+// maximum number of indexers per ID and an overall lease limit. This is useful
+// to ensure the pool does not grow too large, even if some IDs are slow to
+// release indexers.
+type BulkIndexerPool struct {
+	indexers chan *BulkIndexer
+	entries  map[string]idEntry
+	mu       sync.RWMutex
+	cond     *sync.Cond   // To wait/signal when a slot is available.
+	leased   atomic.Int64 // Total number of leased indexers across all IDs.
+
+	// Read only fields.
+	min, max, total int64
+	config          BulkIndexerConfig
+}
+
+type idEntry struct {
+	nonEmpty chan *BulkIndexer
+	count    *atomic.Int64
+}
+
+// NewBulkIndexerPool returns a new BulkIndexerPool with:
+// - The specified guaranteed indexers per ID
+// - The maximum number of concurrent BulkIndexers per ID
+// - A total (max) number of indexers to be leased per pool.
+// - The BulkIndexerConfig to use when creating new indexers.
+func NewBulkIndexerPool(guaranteed, max, total int, c BulkIndexerConfig) *BulkIndexerPool {
+	p := BulkIndexerPool{
+		indexers: make(chan *BulkIndexer, total),
+		entries:  make(map[string]idEntry),
+		min:      int64(guaranteed),
+		max:      int64(max),
+		total:    int64(total),
+		config:   c,
+	}
+	p.cond = sync.NewCond(&p.mu)
+	return &p
+}
+
+// Get returns a BulkIndexer for the specified ID. If there aren't any existing
+// indexers for the ID, a new one is created.
+//
+// If the overall limit of indexers has been reached, it will wait until a slot
+// is available. This is a blocking operation.
+func (p *BulkIndexerPool) Get(id string) *BulkIndexer {
+	// Acquire the lock to ensure that we are not racing with other goroutines
+	// that may be trying to acquire or release indexers.
+	// Even though this looks like it would prevent other Get() operations from
+	// proceeding, the lock is only held for a short time while we check the
+	// count and overall limits. The lock is released by:
+	// - p.cond.Wait() releases the lock while waiting for a signal.
+	// - p.mu.Unlock() releases the lock after the indexer is returned.
+	p.mu.Lock()
+	entry, exists := p.entries[id]
+	if !exists {
+		defer p.mu.Unlock()
```

Comment:
The distance between the locking here and unlocking in `p.get` worries me. Could we maybe refactor this a bit, to something like:

```go
func (p *BulkIndexerPool) Get(id string) *BulkIndexer {
	p.claim() // does the mutex/cond/atomics dance, but does not return a bulk indexer
	return p.getAvailable() // does everything get does now except the atomics & unlock
}
```

That does mean `Get` is slightly more expensive in the case where an entry doesn't yet exist, but that's only going to happen once, so I think the simplification would be worthwhile.

Reply:
I think we can just defer the `p.mu.Unlock`, then we don't need to refactor anything.